onnxscript/function_libs/torch_aten/graph_building.py

```diff
@@ -0,0 +1,446 @@
+    for key, attr_proto in attr_name_to_protos.items():
+        if key not in onnx_attributes:
+            onnx_attributes[key] = attr_proto.value
+
+    return onnx_inputs, onnx_attributes
+
+
+@beartype
+def _unwrap_tensor_to_torch_value(
+    value: ValidArgumentType
+    | dict[str, ValidArgumentType]
+    | list[ValidArgumentType]
+    | tuple[ValidArgumentType, ...]
+) -> torch.Value | str | int | float | dict[str, torch.Value | str | int | float] | list[
+    torch.Value | str | int | float
+] | tuple[torch.Value | str | int | float, ...]:
+    """Unwrap the TorchScriptTensor to torch.Value."""
+    if isinstance(value, dict):
+        return {
+            k: v.symbolic_value() if isinstance(v, TorchScriptTensor) else v
+            for k, v in value.items()
+        }
+    if isinstance(value, list):
+        return [v.symbolic_value() if isinstance(v, TorchScriptTensor) else v for v in value]
+    if isinstance(value, tuple):
+        return tuple(
+            v.symbolic_value() if isinstance(v, TorchScriptTensor) else v for v in value
+        )
+    if isinstance(value, TorchScriptTensor):
+        return value.symbolic_value()
+    # A normal python value
+    return value
+
+
+@beartype
+def _wrap_torch_value_to_tensor(
+    value: torch.Value
+    | dict[str, torch.Value | str | int | float]
+    | list[torch.Value | str | int | float]
+    | tuple[torch.Value | str | int | float, ...]
+) -> ValidArgumentType | dict[str, ValidArgumentType] | list[ValidArgumentType] | tuple[
+    ValidArgumentType, ...
+]:
+    """Wrap torch.Value to TorchScriptTensor."""
+    if isinstance(value, dict):
+        return {
+            k: TorchScriptTensor(v) if isinstance(v, torch.Value) else v
+            for k, v in value.items()
+        }
+    if isinstance(value, list):
+        return [TorchScriptTensor(v) if isinstance(v, torch.Value) else v for v in value]
+    if isinstance(value, tuple):
+        return tuple(TorchScriptTensor(v) if isinstance(v, torch.Value) else v for v in value)
+    if isinstance(value, torch.Value):
+        return TorchScriptTensor(value)
+    return value
+
+
+def _unwrap_tensors_to_torch_values(tensors):
+    # TODO(justinchuby): Do we really need this?
+    if isinstance(tensors, Sequence):
+        return [_unwrap_tensor_to_torch_value(output) for output in tensors]
+    return _unwrap_tensor_to_torch_value(tensors)
+
+
+class TorchScriptEvaluator(evaluator.Evaluator):
+    """An onnxscript Evaluator that captures the graph into torchscript."""
+
+    def __init__(self, graph: TorchScriptGraph):
+        self._graph: TorchScriptGraph = graph
+
+    @property
+    def graph(self) -> TorchScriptGraph:
+        return self._graph
+
+    @beartype
+    def eval_function(
+        self,
+        function: onnxscript.OnnxFunction,
+        args: Sequence[ValidArgumentType],
+        kwargs: Sequence[ValidArgumentType],
+    ):
+        # args/kwargs are TorchScriptTensor/python built-in based
```

Comment:
Does it mean when I see
```
z = torch.ops.aten.add.default(x, y, 5)
```

I can call
```
etor = TorchScriptEvaluator(...)
z = etor.eval_function(torch.ops.aten.add.default, x, y)
```
? I am just wondering what do "args" and "kwargs" mean here, and how this is used inside converter.

Reply:
In spirit yes; in reality some wrapping needs to be done:

When x, y are symbolic tensors of type `TorchScriptTensor`:

```python
evaluator = TorchScriptEvaluator(graph)
with onnxscript.evaluator.default_as(evaluator):
    z = atenlib.aten_add(x, y, 5)
```

When x, y are np arrays:

```python
# This is using ORT by default
z = atenlib.aten_add(x, y, 5)
```